utils/s3/client.cc

```diff
@@ -1092,6 +1093,124 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece, as));
 }
 
+class client::chunked_download_source final : public seastar::data_source_impl {
+    struct claimed_buffer {
+        temporary_buffer<char> _buffer;
+        semaphore_units<> _claimed_memory;
+        s3_clock::duration _consumption_time;
+        claimed_buffer(temporary_buffer<char>&& buf, semaphore_units<>&& claimed_memory, s3_clock::duration consumption_time)
+            : _buffer(std::move(buf)), _claimed_memory(std::move(claimed_memory)), _consumption_time(consumption_time) {}
+    };
+    shared_ptr<client> _client;
+    sstring _object_name;
+    seastar::abort_source* _as;
+    range _range;
+    static constexpr size_t _max_buffers_size = 5_MiB;
+    std::deque<claimed_buffer> _buffers;
+    size_t _buffers_size = 0;
+    bool _is_closing = false;
+    seastar::condition_variable _buffer_cv;
+    future<> _filling_fiber = make_ready_future<>();
+
+    future<> make_filling_fiber() {
+        while (true) {
+            try {
+                s3l.trace("Fiber starts cycle for object '{}'", _object_name);
+                if (_is_closing) {
+                    s3l.trace("Abandoning fiber for object '{}'", _object_name);
+                    co_return;
+                }
+                if (_buffers_size >= _max_buffers_size) {
+                    try {
+                        // just in case we miss the wake-up call use timed `when`
```

Comment:
How come can we "miss" it? Saving own arses with timeout doesn't sound like a good plan

Reply:
`wait` checks the `_signalled` flag and then subscribes the `awaiter` to be awaiken in the case of `broadcast`. Then what can happen, the fiber fills some buffers, gives up with `yield`, `get()` kicks in, takes the buffer and calls `broadcast` which iterates over awaiters and signals each of them, at this moment the fiber's `when` is not subscribed yet, so we may never wake up.